### DISCUSSION TITLE
Allow multiple triplets (primary and fallback for eg x64-windows-static-md and x64-windows) in build system integration via find_package

### DISCUSSION THREAD

**Discussion by ankurvdev** [microsoft/vcpkg#26901] [Ideas]
****

**Is your feature request related to a problem? Please describe.**
Allow  mix and match 2 triplets (for eg x64-windows-static-md and x64-windows) in a single environment
There's a use case where I need to link with a few ports statically and others dynamically (because they dont support static-linkages)

The problem however is VCPKG doesnt understand multiple triplets . 

I can only specify one triplet in VCPKG_TRIPLET to have it look for packages in.

**Proposed solution**

Allow VCPKG_TRIPLET  to be a list. The first is the primary triplet and if a port isnt found in the primary triplet. It can fallback to search the next one in the list.

**Describe alternatives you've considered**

Current workarounds rely upon manually searching using find_library and find_paths for headers and libs and manually setting up the cmake targets in case we need binaries and packages from the non-primary triplet

**Additional context**

**Comment by Osyotr** [comment]
https://github.com/microsoft/vcpkg/blob/c175d8efa7de016dc7a8a22b5c76d201db3ef2d5/docs/users/triplets.md#per-port-customization

Also `vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)` can be used inside a port.

**Comment by ankurvdev** [comment]
The documentation seems to describe how to create a customized triplet.
I'm not sure if thats entirely relevant to the issue i'm raising.
To clarify here's how one would use vcpkg ports in a project

```cmake
include(${VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake)
```
OR
```cmake
set(CMAKE_TOOLCHAIN_FILE "${CMAKE_CURRENT_SOURCE_DIR}/vcpkg/scripts/buildsystems/vcpkg.cmake"
  CACHE STRING "Vcpkg toolchain file")
```
And by default vcpkg will select a triplet for you

Now if you want to switch to an alternate triplet which is non-default
You'd do something like

```cmake
set(VCPKG_TARGET_TRIPLET "x64-windows-static-md")
include(${VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake)
```

And after that

`find_package(foo)` automatically picks up the binaries from `x64-windows-static-md` triplet 

However if you want both 
```
find_package(foo)
find_package(bar)
```
but we want it to pick up `foo:x64-windows-static-md` but `bar:x64-windows` because I want to link dynamically with `bar` but statically with `foo` the vcpkg cmake integration doesnt allow for that.

So instead I'm proposing here that instead of  `VCPKG_TARGET_TRIPLET` being a single string, we make it a list so i can specify the preference order 
for eg:
```
set(VCPKG_TARGET_TRIPLET "x64-windows-static-md;x64-windows")
include(${VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake)
find_package(foo)
find_package(bar)
```

and as long as the `VCPKG_ROOT` contains `foo::x64-windows-static-md` and not `bar::x64-windows-static-md`,  `bar:x64-windows` is picked up because the `find_package` falls back to the second entry in `VCPKG_TARGET_TRIPLET`

**Reply by dg0yt** [reply]
> `bar:x64-windows` is picked up because the `find_package` falls back to the second entry in `VCPKG_TARGET_TRIPLET`

... until `bar:x64-windows-static-md` suddenly becomes present, because you installed it by accident, or because it becomes a dependency of another `x64-windows-static-md` port. This is not reliable. If you want a mixed universe, you must explicitly define this universe - by a customized triplet file.
Not to mention manifest mode.

**Reply by ankurvdev** [reply]
I agree its not reliable for the general case but there's a huge set of case where there's no dependencies between foo and bar where this could work 

Also,  I don't think there are alternatives.
I am not aware of how I can create a customized triplet that would let me install both 
`foo:mytriplet`
and 
`bar:mytriplet`
but foo is built statically and bar dynamically.

How would this customized triplet work? Can you please elaborate ?

**Reply by dg0yt** [reply]
The documentation was linked in the first answer, but again:
https://github.com/microsoft/vcpkg/blob/c175d8efa7de016dc7a8a22b5c76d201db3ef2d5/docs/users/triplets.md#per-port-customization

**Reply by ankurvdev** [reply]
Ah ok ... Sorry I missed the context earlier, that the triplet file itself contains ports that we want dynamically linked.

Thanks for reposting. I think i understand now.
